mountpoint-s3/src/inode.rs

```diff
@@ -678,8 +678,16 @@ struct InodeInner {
     full_key: String,
     kind: InodeKind,
 
-    // Mutable inode state. This lock should also be used to serialize operations on an inode (like
-    // creating a new child).
+    /// Mutable inode state. This lock should also be held to serialize operations on an inode (like
+    /// creating a new child).
+    ///
+    /// When taking a lock across multiple [Inode]s,
+    /// we must always acquire the locks in the following order to avoid deadlock:
+    ///
+    /// - Any ancestors in descending order, if they need to be locked.
+    /// - Otherwise, ascending order by [InodeNo].
+    ///   This reflects similar behavior in the Kernel's VFS named 'inode pointer order',
+    ///   described in https://www.kernel.org/doc/html/next/filesystems/directory-locking.html
```

Comment:
It was great to nice to add rustdoc for the rules to follow. I was still not aware of `Otherwise, ascending order by [InodeNo].`